# Домашнее задание 
# Математика — Домашнее задание 9
# А.&#8201;Д. Егоров<h1/>

## Условие задачи:
Задача максимизации:
\begin{equation*}
    L = 2x_1 + 3x_2 - x_4 \rightarrow max,
\end{equation*}
ограничения:
\begin{equation*}
    \begin{cases}
        2x_1 - x_2 -2x_4 + x_5 = 16, \\
        3x_1 + 2x_2 + x_3 - 3x_4 =18, \\
        -x_1 + 3x_2 + x_4 + x_6 = 24, \\
        x_i \geqslant 0,\ i = \overline{1, 6}.
    \end{cases}
\end{equation*}


## Аналитическое решение 



### Замечание
Заметим, что исходная задача аналогична второму шагу решения следующей задачи:
\begin{equation*}
    L = 2x_1 + 3x_2 - x_4 \rightarrow max,
\end{equation*}
ограничения:
\begin{equation*}
    \begin{cases}
        2x_1 - x_2 -2x_4 \leqslant 16, \\
        3x_1 + 2x_2 - 3x_4 \leqslant 18, \\
        -x_1 + 3x_2 + x_4 \leqslant 24, \\
        x_i \geqslant 0,\ i = 1,\,2,\,4.
    \end{cases}
\end{equation*}

### Тогда для исходной задачи
составим симплекс таблицу, где за базис выберем переменные, не входящие в минимизируемый функционал.

|      *Базис* | $x_1$ | $x_2$ | $x_3$ | $x_4$ | $x_5$ | $x_6$ | *Свободные члены* | *Симплекс-отношения* |
|          --- |   --- |   --- |   --- |   --- |   --- |   --- |               --- |                  --- |
|        $x_3$ |     2 | **-1**|     0 |    -2 |     1 |     0 |                16 |                  -16 |
|        $x_5$ |     3 |  **2**|     1 |    -3 |     0 |     0 |                18 |                   9  |
|        $x_6$ |    -1 |  **3**|     0 |     4 |     0 |     1 |                24 |                   8  |
|            L |    -2 | **-3**|     0 |     1 |     0 |     0 |                 - |                    - |

|      *Базис* |  $x_1$ | $x_2$ | $x_3$ | $x_4$ | $x_5$ | $x_6$ | *Свободные члены* | *Симплекс-отношения* |
|          --- |    --- |   --- |   --- |   --- |   --- |   --- |               --- |                  --- |
|        $x_3$ |**5/3** |     0 |     0 |  -2/3 |     1 |   1/3 |                24 |                 72/5 |
|        $x_5$ |**11/3**|     0 |     1 | -17/3 |     0 |  -2/3 |                 2 |                  6/11|
|        $x_2$ |**-1/3**|     1 |     0 |   4/3 |     0 |   1/3 |                 8 |                   24 |
|            L | **-3** |     0 |     0 |     5 |     0 |     1 |                 - |                    - |

|      *Базис* |  $x_1$ | $x_2$ | $x_3$ | $x_4$ | $x_5$ | $x_6$ | *Свободные члены* | *Симплекс-отношения* |
|          --- |    --- |   --- |   --- |   --- |   --- |   --- |               --- |                  --- |
|        $x_3$ |      0 |     0 | -5/11 | 21/11 |     1 |  7/11 |            254/11 |                    - |
|        $x_1$ |      1 |     0 |  3/11 |-17/11 |     0 | -2/11 |              6/11 |                    - |
|        $x_2$ |      0 |     1 |  1/11 |  9/11 |     0 |  3/11 |             90/11 |                    - |
|            L |      0 |     0 |  9/11 |  4/11 |     0 |  5/11 |                 - |                    - |

Все коэффициенты при $L$ стали положительными, следовательно, задача решена. Запишем ответ: 

Оптимальный план: 
$$
    X = \left( \dfrac{6}{11},\ \dfrac{90}{11},\ 0,\ 0,\ \dfrac{254}{11},\ 0  \right)^T
$$

Максимум функционала: 
$$
    L_{max} = 2 \cdot \dfrac{6}{11} + 3 \cdot \dfrac{90}{11} - 1 \cdot 0 = \dfrac{282}{11} \approx 25.6364
$$

## Численное решение

In [1]:
import numpy as np
import scipy.optimize as opt

#### С использованием `scipy`

|      *Базис* | $x_1$ | $x_2$ | $x_3$ | $x_4$ | $x_5$ | $x_6$ | *Свободные члены* | *Симплекс-отношения* |
|          --- |   --- |   --- |   --- |   --- |   --- |   --- |               --- |                  --- |
|        $x_3$ |     2 | **-1**|     0 |    -2 |     1 |     0 |                16 |                  -16 |
|        $x_5$ |     3 |  **2**|     1 |    -3 |     0 |     0 |                18 |                   9  |
|        $x_6$ |    -1 |  **3**|     0 |     4 |     0 |     1 |                24 |                   8  |
|            L |    -2 | **-3**|     0 |     1 |     0 |     0 |                 - |                    - |

In [2]:
c = [2, 3, 0, -1, 0, 0]
c = [-x for x in c] #for maximization
print(c)
A = [[2, -1, 0, -2, 1, 0], [3, 2, 1, -3, 0, 0], [-1, 3, 0, 4, 0, 1]]
b = [16, 18, 24]

opt.linprog(c=c, A_eq=A, b_eq=b)

[-2, -3, 0, 1, 0, 0]


        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -25.636363636363637
              x: [ 5.455e-01  8.182e+00  0.000e+00  0.000e+00  2.309e+01
                   0.000e+00]
            nit: 3
          lower:  residual: [ 5.455e-01  8.182e+00  0.000e+00  0.000e+00
                              2.309e+01  0.000e+00]
                 marginals: [ 0.000e+00  0.000e+00  8.182e-01  3.636e-01
                              0.000e+00  4.545e-01]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
          eqlin:  residual: [ 0.000e+00  0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -8.182e-01 -4.545e-01]
        ineqlin:  residual: []
                 marginals: []
 mip_node_count: 0
 mip_dual_bo

Ответ `fun = -25.636363636363637` надо взять противоположного знака, т.к. по сути здесь мы минимизировали исходный функционал, домноженный на $-1$. 


Итого, ответ `scipy`: $L_{max} = 25.636363636363637$

#### Своя реализация метода

In [3]:
def make_simplex_table(c, c0, A, b):
    c0 = c0 if c0 != None else 0
    n, m = A.shape
    basis = np.arange(n) + m
    row_constraints = np.hstack((A, np.eye(n), b[:, np.newaxis]))
    row_objective = np.hstack((c, np.zeros(n), c0))
    row_pseudo_objective = -row_constraints.sum(axis=0)
    row_pseudo_objective[basis] = 0

    return np.vstack((row_constraints, row_objective, row_pseudo_objective))


def get_pivot_col(table, tol=1e-9):
    # Non negative condition
    # if True => solution found
    if np.where(table[-1, :-1] <= -tol)[0].shape[0] == 0:
        return False, np.nan

    pivot_col = np.where(table[-1, :-1] == np.min(table[-1, :-1]))[0][0]

    return True, pivot_col


def get_pivot_row(table, basis, pivot_col, phase, tol=1e-9):
    if phase == 0:
        k = 2
    else:
        k = 1

    # Avoid division by zero
    # just divide by very small number
    theta_den = np.where(
        np.abs(table[:-k, pivot_col]) > tol,
        table[:-k, pivot_col],
        np.full(table.shape[0] - k, 1e-9)
    )

    theta = table[:-k, -1] / theta_den
    theta_above_zero = np.where(theta >= tol)[0]
    if theta_above_zero.shape[0] == 0:
        return False, np.nan
    pivot_row = np.where(theta == theta[theta_above_zero].min())[0][0]
    return True, pivot_row


def recalculate_table(table, basis, row_min, col_min, tol):
    basis[row_min] = col_min
    leader = table[row_min, col_min]
    table[row_min] = table[row_min] / leader
    for i in range(table.shape[0]):
        if i != row_min:
            table[i] = table[i] - table[i, col_min] * table[row_min]


def solve_simplex(table, n, basis, phase=0, maxiter=1000, tol=1e-9, nit_phase_0=0):
    nit = nit_phase_0
    compete = False
    while not compete:
        pivot_col_found, col_min = get_pivot_col(table, tol)
        if not pivot_col_found:
            compete = True
            break
        else:
            pivot_row_found, row_min = get_pivot_row(table, basis, col_min, phase, tol)
            if not pivot_row_found:
                compete = True
                raise Exception("Task has no solution")

        if nit >= maxiter:
            compete = True
        else:
            recalculate_table(table, basis, row_min, col_min, tol)
            nit += 1

    return nit


def my_simplex_impl(c, c0=None, A=None, b=None, maxiter=1000, tol=1e-9):
    """
    Simplex method implementation
    method="artificial basis"

    c : functional coefs.
    c0 : free functional coef.
    A : matrix with conditions.
    b : free coeficients of conditions.
    maxiter: limit for iterations.
    tol: tolerance.
    """
    A = np.array(A)
    b = np.array(b)
    c = np.array(c)

    is_negative_constraint = np.less(b, 0)
    A[is_negative_constraint] *= -1
    b[is_negative_constraint] *= -1

    n, m = A.shape
    # Get basis
    basis = np.arange(n) + m
    basis_first = basis.copy()

    # Get simplex matrix for artificial basis method
    table = make_simplex_table(c, c0, A, b)

    # phase 0 (with pseudo objective row)
    nit_phase_0 = solve_simplex(table, n, basis, phase=0, tol=tol)

    # if pseudo objective row zero remove it from tableau
    if abs(table[-1, -1]) < tol:
        table = table[:-1, :]
        table = np.delete(table, basis_first, 1)
    else:
        raise Exception("Task cannot be solved")

    # phase 1 (no pseudo objective row)
    nit = solve_simplex(table, n, basis, phase=1, tol=tol, nit_phase_0=nit_phase_0)

    # optimal x
    x = np.zeros(m)
    for i in range(n):
        x[basis[i]] = table[i, -1]

    # functional optimum
    fun = sum([c[i] * x[i] for i in range(len(x))]) + (c0 if c0 != None else 0)
    return x, fun, nit

In [4]:
c = [2, 3, 0, -1, 0, 0]
c = [-x for x in c]  # for maximization
A = [[2, -1, 0, -2, 1, 0], [3, 2, 1, -3, 0, 0], [-1, 3, 0, 4, 0, 1]]
b = [16, 18, 24]

x, fun, _ = my_simplex_impl(c=c, A=A, b=b) 
res = f" opt_x = {x} \n L(opt_x) = {fun}"
print(res)

 opt_x = [ 0.54545455  8.18181818  0.          0.         23.09090909  0.        ] 
 L(opt_x) = -25.636363636363637


Ответ `fun = -25.636363636363637` надо взять противоположного знака, т.к. по сути здесь мы минимизировали исходный функционал, домноженный на $-1$. 


Итого, ответ: $L_{max} = 25.636363636363637$